### ISSUE TITLE
@tenhtof but we don't have transformRequest props for MapView component

### ISSUE THREAD

**Issue by dsardar099** [maplibre/maplibre-react-native#424]
@tenhtof but we don't have transformRequest props for MapView component

_Originally posted by @dsardar099 in https://github.com/maplibre/maplibre-react-native/issues/415#issuecomment-2231063304_

**Comment by tyrauber** [comment]
@tenhtof but we don't have transformRequest props for MapView component

_Originally posted by @dsardar099 in https://github.com/maplibre/maplibre-react-native/issues/415#issuecomment-2231063304_

**Comment by ngottlieb** [comment]
It'd be great if this can include `transformStyle` as well if and when someone decides to tackle it!

**Comment by viettranbhsoft1** [comment]
Any update for this?

**Comment by edurenye** [comment]
To connect to my TileServer GL, I had to use a proxy to convert a URL encoded token into a header token, as a workaround.
But would be great to be able to alter the request so I can add a header without needing a proxy server for it. Since this work around is quite expensive.

**Comment by tyrauber** [comment]
@edurenye, I believe you can address this usecase with [addCustomHeader](https://maplibre.org/maplibre-react-native/docs/guides/custom-http-headers#adding-a-header).

**Comment by edurenye** [comment]
Thank you very much @tyrauber !

**Comment by 6TELOIV** [comment]
I've been running into this same issue (with Esri TileJSON being non spec compliant), and I'd like to contribute this feature (though I'm not sure I'd be able to).

Some questions for the maintainers:
- Would this require changes upstream in [maplibre-native](https://github.com/maplibre/maplibre-native)? It seems to me like it would, as this library doesn't handle the requests.
- If upstream changes are needed, how do y'all coordinate that work across multiple projects? It seems like [there's a desire for this feature over there as well](https://github.com/maplibre/maplibre-native/issues/126), but the issue is stale and hasn't had any maintainer input.

**Comment by KiwiKilian** [comment]
MLRN is only a wrapper around MLN. So if something is not possible within MLN, MLRN can't solve it. I don't fully understand the limitations over in https://github.com/maplibre/maplibre-native/issues/126, but one would have to try, if this could be used to implement the desired feature. If not, you can try to implement it within MLN, then adopting it here for MLRN.

My personal opinion: Instead of spending time on implementing this, make your styles spec compliant.

**Comment by 6TELOIV** [comment]
@KiwiKilian I agree wholeheartedly! However, I and other Esri users don't control what their API returns. I have let them know about the issue, but it's been known since at least 2021 so IDK if anything will come of it.

I think the approach outlined in https://github.com/maplibre/maplibre-native/issues/126 won't support the transformRequest feature of maplibre-gl-js as it won't have context about what type of request it is.

**Comment by KiwiKilian** [comment]
While we can't support the exact same API like `transformRequest` we do now have the [`TransformRequestManager`](https://github.com/maplibre/maplibre-react-native/blob/beta/package/src/modules/transform-request/TransformRequestManager.ts) in the latest beta.  It should support most common URL rewriting cases.

Why can't we implement `transformRequest`?

AFAIK this is simply not possible through plain React Native, to have a callback returning a value to native. We would have to switch to Nitro Modules, which [support this](https://nitro.margelo.com/docs/types/callbacks#callbacks-that-return-a-value---t). This is currently out of scope.
